### PR TITLE
feat!: Replace `sources` with `contentDirectories`

### PR DESCRIPTION
**Pull request by akash1810**
Fixes https://github.com/guardian/actions-riff-raff/issues/13.

> **Note**
> This will need a new major version.
> Updates to the README can be more easily viewed [here](https://github.com/guardian/actions-riff-raff/blob/aa-sources/README.md).

## What does this change?

In this change we remove support for `sources` in the body of the `riff-raff.yaml` config, and replace it with a `contentDirectories` input to the Action.

This decoupling keeps the `riff-raff.yaml` config "valid". That is, we're able to copy/paste it to [Riff-Raff's validation tool](https://riffraff.gutools.co.uk/configuration/validation) w/out first needing to remove `sources`.

It also means this action becomes compatible with auto-generated `riff-raff.yaml` files.

⚠️  BREAKING CHANGE: `sources` within `riff-raff.yaml` (either via the `config` or `configPath` input) is no longer supported

Configure the files to upload via the `contentDirectories` input, where `contentDirectories` is a mapping of package name -> artifact files.

From:

```yaml
- uses: guardian/actions-riff-raff@v1
  with:
    app: my-application
    config: |
      stacks:
        - deploy
      regions:
        - eu-west-1
      allowedStages:
        - CODE
        - PROD
      deployments:
        static-site-assets:
          type: aws-s3
          sources:
            - static-site/dist # upload the `dist` directory
          parameters:
            bucket: aws-some-bucket
            cacheControl: private
            publicReadAcl: false
        api:
          type: autoscaling
          sources:
            - target/application.jar
          parameters:
            bucketSsmLookup: true
```

To:

```yaml
- uses: guardian/actions-riff-raff@v2 # Note the version change
  with:
    app: my-application
    config: |
      stacks:
        - deploy
      regions:
        - eu-west-1
      allowedStages:
        - CODE
        - PROD
      deployments:
        static-site-assets: # <-- this is a package name
          type: aws-s3
          parameters:
            bucket: aws-some-bucket
            cacheControl: private
            publicReadAcl: false
        api: # <-- this is another package name
          type: autoscaling
          parameters:
            bucketSsmLookup: true
    contentDirectories: |
      static-site-assets:
        - static-site/dist
      api:
        - target/application.jar
```

## How to test

See:
  - https://github.com/guardian/cdk-playground/pull/284
  - https://github.com/guardian/tracker/pull/551

### CODE REVIEW

**Review by nicl — COMMENTED** [review]
Have left some thoughts on testing especially - as would be keen for us to avoid the `test-utils` approach here. But in terms of feature-set this is great!

File: README.md [hunk 1]
```diff
@@ -36,6 +36,22 @@ To use, add (something like) the following to your workflow file:
             publicReadAcl: false
 ```
 
+Or, to use with an existing `riff-raff.yaml`:
```
[the remark this answers is not in the document]
Comment (akash1810) [reply README.md:39]: > (I would recommend we bias towards inlining here as for new apps that is the simpler approach - it avoids another file and keeps things co-located.)

I think I disagree with this. I do not think we should support inlined config at all as: new apps should use GuCDK and then generate the config; existing apps should already have a config file.

Whilst the co-location is nice, there are some (albeit not too big) advantages to a standalone file, namely the IDE helps you format it (YAML is sensitive to whitespace!).

If we also remove reading `sources` from the config (as in #19), I think the DX to adopt this action is simplified:

```
Add this to your .github/workflows/ci.yaml file:

- uses: guardian/actions-riff-raff@v1
  with:
    app: foo
    configPath: /path/to/riff-raff.yaml
    contentDirectories: |
      - my-cloudformation:
        - cfn/template.yaml
      - my-lambda-app:
        - lambda.zip
```

File: src/test-utils.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import * as yaml from 'js-yaml';
+
+// Read yaml config and set env vars to mimic GHA
+export const readConfig = (yamlConfig: string): void => {
```
[the remark this answers is not in the document]
Comment (akash1810) [reply src/test-utils.ts:4]: This makes complete sense 👍🏽 .

However, I think there's still an underlying issue of state persisting across tests. We have 3 tests, each sets some env vars via this `readConfig` function. As env vars persist across tests, it's necessary to remove them (either in `beforeEach` or `afterEach`), hence `clearConfig`.

As the tests that use `readConfig` and `clearConfig` sit across multiple files, the functions were moved to this utils file. An alternative would have been to merge the tests into a single file, which I don't think is a good idea.

Not sure what a good solution is here...

File: README.md [hunk 3]
```diff
@@ -1,10 +1,10 @@
-# actions-riff-raff
+# guardian/actions-riff-raff
```
[the remark this answers is not in the document]
Comment (akash1810) [reply README.md:1]: Don't think so - `@` is for NPM packages, when we add this Action to a GH Workflow, it's `- uses: guardian/actions-riff-raff` (`- uses: <org>/<repo>`).

File: src/test-utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (akash1810) [reply src/test-utils.ts:4]: I've thought about this some more and think we can completely drop the tests for config, bringing us back to the previous situation of the single integration test of index.

The config tests were added purely to test the behaviour of `sources` and `contentDirectories`. In #19 we remove `sources` entirely - if we make that change in this PR, the need for the test disappears. I'll do this now, and close #19.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (akash1810) [reply README.md]: The README has been tidied up now - I'll resolve this comment as it's stale. I'd still be keen to explore dropping the `config` input, but that isn't tied to this change, and would also need more discussion.

**Review by nicl — APPROVED** [review]
LGTM! I'm tempted to attempt a slight shortening of the README as a follow up but this is a big improvement on things as it stands.

**Pull request merged**